programs/mpl-core/src/processor/update_external_plugin.rs

```diff
@@ -16,24 +30,273 @@ pub(crate) struct UpdateExternalPluginV1Args {
+    }
+
+    let (_, plugin) =
+        fetch_wrapped_external_plugin::<AssetV1>(ctx.accounts.asset, None, &args.key)?;
+
+    let (mut asset, plugin_header, plugin_registry) = validate_asset_permissions(
+        authority,
+        ctx.accounts.asset,
+        ctx.accounts.collection,
+        None,
+        None,
+        Some(&plugin),
+        AssetV1::check_update_external_plugin,
+        CollectionV1::check_update_external_plugin,
+        PluginType::check_update_external_plugin,
+        AssetV1::validate_update_external_plugin,
+        CollectionV1::validate_update_external_plugin,
+        Plugin::validate_update_external_plugin,
+        None,
+        None,
+    )?;
+
+    let mut plugin_registry = plugin_registry.ok_or(MplCoreError::PluginsNotInitialized)?;
+    let mut plugin_header = plugin_header.ok_or(MplCoreError::PluginsNotInitialized)?;
+
+    let plugin_registry_clone = plugin_registry.clone();
+    let (_, record) = find_external_plugin(&plugin_registry_clone, &args.key, ctx.accounts.asset)?;
+    let mut registry_record = record.ok_or(MplCoreError::PluginNotFound)?.clone();
+    registry_record.update(&args.update_info);
+
+    let mut new_plugin = plugin.clone();
+    new_plugin.update(&args.update_info);
+
+    let plugin_data = plugin.try_to_vec()?;
+    let new_plugin_data = new_plugin.try_to_vec()?;
+
+    // The difference in size between the new and old account which is used to calculate the new size of the account.
+    let plugin_size = plugin_data.len() as isize;
+    let size_diff = (new_plugin_data.len() as isize)
+        .checked_sub(plugin_size)
+        .ok_or(MplCoreError::NumericalOverflow)?;
+
+    // The new size of the account.
+    let new_size = (ctx.accounts.asset.data_len() as isize)
+        .checked_add(size_diff)
+        .ok_or(MplCoreError::NumericalOverflow)?;
+
+    // The new offset of the plugin registry is the old offset plus the size difference.
+    let registry_offset = plugin_header.plugin_registry_offset;
+    let new_registry_offset = (registry_offset as isize)
+        .checked_add(size_diff)
+        .ok_or(MplCoreError::NumericalOverflow)?;
+    plugin_header.plugin_registry_offset = new_registry_offset as usize;
+
+    // The offset of the first plugin is the plugin offset plus the size of the plugin.
+    let next_plugin_offset = (registry_record.offset as isize)
+        .checked_add(plugin_size)
+        .ok_or(MplCoreError::NumericalOverflow)?;
+
+    let new_next_plugin_offset = next_plugin_offset
+        .checked_add(size_diff)
+        .ok_or(MplCoreError::NumericalOverflow)?;
+
+    // //TODO: This is memory intensive, we should use memmove instead probably.
+    let src =
+        ctx.accounts.asset.data.borrow()[(next_plugin_offset as usize)..registry_offset].to_vec();
+
+    resize_or_reallocate_account(
+        ctx.accounts.asset,
+        ctx.accounts.payer,
+        ctx.accounts.system_program,
+        new_size as usize,
+    )?;
```

Comment:
Would it be worth putting a check around this (and in the other update functions) like `if plugin_size != new_size` so it can be skipped if the size is same?  Not sure if there would be a meaningful compute difference.

Reply:
Yes this is a good call. Honestly I could just put this in the resize function itself for a global optimization.